src/main/java/org/opensearch/neuralsearch/processor/TextEmbeddingProcessor.java

```diff
@@ -0,0 +1,278 @@
+import org.opensearch.action.ActionListener;
+import org.opensearch.client.Client;
+import org.opensearch.common.CheckedConsumer;
+import org.opensearch.ingest.AbstractProcessor;
+import org.opensearch.ingest.IngestDocument;
+import org.opensearch.ingest.Processor;
+import org.opensearch.ml.client.MachineLearningNodeClient;
+import org.opensearch.neuralsearch.ml.MLCommonsClientAccessor;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+
+@Log4j2
+public class TextEmbeddingProcessor extends AbstractProcessor {
+
+    public static final String TYPE = "text_embedding";
+    public static final String MODEL_ID_FIELD = "model_id";
+    public static final String FIELD_MAP_FIELD = "field_map";
+
+    private static final String LIST_TYPE_NESTED_MAP_KEY = "knn";
+
+    private final String modelId;
+
+    private final Map<String, Object> fieldMap;
+
+    private final MLCommonsClientAccessor mlCommonsClientAccessor;
+
+    public TextEmbeddingProcessor(String tag, String description, String modelId, Map<String, Object> fieldMap, Client client) {
+        super(tag, description);
+        this.modelId = Objects.requireNonNull(modelId, "model_id is null, can not process it");
+        if (fieldMap == null || fieldMap.size() == 0 || checkEmbeddingConfigNotValid(fieldMap)) {
+            throw new IllegalArgumentException("filed_map is null, can not process it");
+        } else {
+            this.fieldMap = fieldMap;
+        }
+        this.mlCommonsClientAccessor = new MLCommonsClientAccessor(new MachineLearningNodeClient(client));
+    }
+
+    private boolean checkEmbeddingConfigNotValid(Map<String, Object> fieldMap) {
+        return fieldMap.entrySet().stream().anyMatch(x -> StringUtils.isBlank(x.getKey()) || Objects.isNull(x.getValue()));
+    }
+
+    @Override
+    public IngestDocument execute(IngestDocument ingestDocument) {
+        validateEmbeddingFieldsType(ingestDocument, fieldMap);
+        Map<String, Object> knnMap = buildKnnMap(ingestDocument, fieldMap);
+
+        ActionListener<List<List<Float>>> internalListener = ActionListener.wrap(
+            responseConsumer(ingestDocument, knnMap),
+            exceptionConsumer()
+        );
+
+        mlCommonsClientAccessor.inferenceSentences(this.modelId, buildMLInput(knnMap), internalListener);
+        return ingestDocument;
+    }
+
+    @VisibleForTesting
+    CheckedConsumer<List<List<Float>>, Exception> responseConsumer(IngestDocument ingestDocument, Map<String, Object> knnMap) {
+        return res -> {
+            Objects.requireNonNull(res, "embedding failed!");
+            Map<String, Object> vectorOutput = buildVectorOutput(knnMap, res, ingestDocument.getSourceAndMetadata());
+            vectorOutput.forEach(ingestDocument::appendFieldValue);
+        };
+    }
+
+    @VisibleForTesting
+    Consumer<Exception> exceptionConsumer() {
+        return exception -> log.error(exception.getMessage(), exception);
+    }
```

Comment:
User has the flexibility to determine what to do when error arise, please refer to this: https://www.elastic.co/guide/en/elasticsearch/reference/master/ingest.html#handling-pipeline-failures. And a processor itself doesn't have the capability to prevent document ingestion.

Reply:
The use can set the tags of ignore_failure to proceed. But at least we need to announce the failure if they choose not to ignore_failure.

Reply:
Even if we want to announce the failures, we should put a proper message instead of saying exception.getMessage().